gloo/cuda_collectives_device.h

```diff
@@ -32,120 +32,228 @@ class CudaLocalDeviceReduce : public LocalOp<T> {
+  // the target pointer. Then we know the final memcpy can be performed using
+  // peer access as well.
+  template <typename Dst1 = Dst>
+  void alignIndicesWithTarget(
+      const std::vector<std::vector<int>>& groups,
+      typename std::enable_if<
+          std::is_same<Dst1, CudaDevicePointer<T>>::value>::type* = 0) {
+    auto targetGroup = findGroup(groups, targetPtr_.getDeviceID());
+    if (targetGroup != 0) {
+      auto it = indices_.begin() + targetGroup;
+      auto group = std::move(*it);
+      indices_.erase(it);
+      indices_.insert(indices_.begin(), std::move(group));
+    }
+  }
 
-        // Record event on secondary stream
-        CUDA_CHECK(cudaSetDevice(devicePtrs_[indexB].getDeviceID()));
-        CUDA_CHECK(cudaEventRecord(
-                       streamB.getEvent(),
-                       streamB.getStream()));
+  // Ensures that the devices in group 0 are closest to the memory in
+  // the target pointer. The final memcpy will then be as fast as possible.
+  template <typename Dst1 = Dst>
+  void alignIndicesWithTarget(
+      const std::vector<std::vector<int>>& groups,
+      typename std::enable_if<
+          std::is_same<Dst1, CudaHostPointer<T>>::value>::type* = 0) {
+    // TODO
+  }
 
-        // Make primary stream wait for secondary stream.
-        // This ensures any operations on the source pointer
-        // have finished before we start the reduction.
-        CUDA_CHECK(cudaSetDevice(devicePtrs_[indexA].getDeviceID()));
-        CUDA_CHECK(cudaStreamWaitEvent(
-                       streamA.getStream(),
-                       streamB.getEvent(),
-                       0));
+  void computeIndices() {
+    // Add level of indirection so that we can shuffle this instead
+    // of shuffling BOTH the streams and device pointer vectors.
+    // Group indices by peer to peer access.
+    const auto& groups = groupFullyConnected();
+    GLOO_ENFORCE_LE(groups.size(), 2, "Expected <= 2 peer access groups");
+    indices_.resize(groups.size());
+    for (auto i = 0; i < devicePtrs_.size(); i++) {
+      indices_[findGroup(groups, devicePtrs_[i].getDeviceID())].push_back(i);
+    }
+    // Remove empty groups
+    for (auto it = indices_.begin(); it != indices_.end();) {
+      if (it->empty()) {
+        it = indices_.erase(it);
+      } else {
+        it++;
+      }
+    }
+    // Best effort alignment so that the group in indices[0] is close to target
+    if (indices_.size() > 1) {
+      alignIndicesWithTarget(groups);
+    }
 
-        // Queue reduction
-        fn_->call(
-            devicePtrs_[indexA],
-            devicePtrs_[indexB],
-            devicePtrs_[indexA].getCount(),
-            streamA);
+    // Shuffle order in an attempt to evenly spread work across devices when
+    // dealing with multiple instances of this operation.
+    auto& groupA = indices_[0];
+    std::random_shuffle(groupA.begin(), groupA.end());
```

Comment:
It may be better to explicitly set up non-overlapping communication topologies and iterate through them. Random may not work great because a single oversubscribed link may lead to 2x overall degradation. Can you make a TODO comment in the code?